flink-ml-core/src/main/java/org/apache/flink/ml/common/sharedobjects/SharedObjectsStreamOperator.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.sharedobjects;
+
+/** Interface for all operators that need to access the shared objects. */
+public interface SharedObjectsStreamOperator {
```

Comment:
As I understand, this PR tries to provide an infrastructure for sharing objects among multiple Flink operators, through java static variables. 

To achieve this, it empoys one specific Flink operator for each sharing object as the writer and others operators as the reader. Based on this, the GBDT implementation relies on the Flink events to guarantee the read/write order of each object. 

However, can you explain some other machine learning algorithms that would use `SharedObjects` in the future? And is there a general way that developers can guarantee the order of read/writes is correct? If a reader of an object changes the value of that object, does it still follows the assumption of `SharedObjects`?

There is another possible solution [1] that we put all the computation logic into one operator (i.e., WorkerOperator) and all the computation logic into another operator (i.e., ServerOperator). In this case, we would not need shared objects anymore. Let's have a thorough comparison between these two options.

[1] https://github.com/apache/flink-ml/pull/237

Reply:
Thanks for your valuable comments. You mentioned several issues in your comments, and I will answer them one by one.

Q: "can you explain some other machine learning algorithms that would use SharedObjects in the future?"

One common pattern where SharedObjects can be used is same datasets are needed in operators before and after a `reduce` operator.  Here I list some algorithms (correct me if I made a mistake):
‒  All distributed algorithms based on decision trees: both model and training data are required during the nodes splitting after reducing the intermediate data.
‒ Second-order gradient optimizer, e.g. Newton, L-BFGS: require the use of old gradient data after the reduce operation.
‒ Algorithms that have two rounds of (unmergeable) `reduce` operation in each iteration: GBDT, ALS, GMM, and LDA.
‒ Evaluation metrics are calculated every few rounds of iterations: evaluation metrics has to be calculated with `reduce` after model is updated.

Q: "is there a general way that developers can guarantee the order of read/writes is correct?"

A simple approach is to connect the reader and writer with an addition dummy stream.  Readers read the data only after receiving elements from the stream.
This approach assumes the reads and writers are not interleaved which should be true in most algorithms. If not (please give some examples), multi-threading techniques like atomics can be used.

Q："If a reader of an object changes the value of that object, does it still follows the assumption of SharedObjects?"

I think this is a common issue in Java, and one solution is to give readers a deep clone of the object, which can be expensive. Balancing efficiency and safety, I chose efficiency.

Reply:
#237 proposes a brilliant solution about abstracting iterative computations inspired from Parameter Server. I believe both solutions can work well in many algorithms/scenarios.

Before comparing two solutions, there are two facts about PS infra I must emphasize:

One fact is the PS infra is built on the DataStream APIs, which means there will be no performance improvement compared to implement with raw DataStream APIs. So we mainly discuss its usability with aspect to developers.

The other fact is the current status of functionalities shown in #237 cannot fully meet the requirements of GBDT implementation.  `MessageType`, model format, reduce logic of messages, etc. are all fixed/hard-coded with respect to gradient-based algorithms. The usability will drop significantly if forcing GBDT implementation to use current APIs.

Therefore, to make a reasonable comparison between two solutions, I assume an extended version of current PS infra which supports POJO message types and POJO model data, user-defined `reduce` function, etc. Here are my thoughts under this assumption:

1. Framework Intrusiveness
Using PS infra means developers cannot use DataStream APIs in iterations anymore. Then there are cases where PS infra cannot implement: 
  - side outputs: evaluation result streams; prediction and model streams in online cases.
  - partition/join/coGroup of training data sets: AUC calculation after model update, ALS, SimRank.

As for SharedObjects, it is an augment to DataStream APIs. There is no extra limitation to developers.

The intrusiveness also influences the observation of operators when job running as, in PS infra, multiple computations are merged in to one operator, like in/out stats, checkpoint status. This decreases usability to both developers and end-users.

2. Applicable scenarios
Besides inapplicable cases mentioned above, PS infra cannot work in non-iteration cases. But SharedObjects can work. One possible case is to improve consecutive joins with a same dataset by reducing a copy of dataset.

3. Learning curve
PS infra provides a whole set of concepts and interfaces such as Message, ModelUpdater, ProcessStage, TrainingUtils, etc., which are not related to the existing DataStream API and have a steeper learning curve.

SharedObjects provides two interfaces, SharedObjectsUtils and SharedObjectsContext, and can be developed directly based on the existing DataStream API code, making it easier for developers to accept.


Overall speaking, I think both solutions can coexist because they are on different levels of APIs and have no conflicts. How about you? @zhipeng93